### PR TITLE
criminal justice analysis data models

### PR DESCRIPTION
**Pull request by actlikewill**
This adds models for the criminal justice dashboards.

### CODE REVIEW

File: peachjam/analysis/criminal_data/offence.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+from .types import MatchOffenceResult, MentionedOffences
+
+log = logging.getLogger(__name__)
+
+
+class OffenceMentionExtractor(BaseExtractor):
+    def build_prompt(self) -> str:
+        judgment_text = self.judgment.get_content_as_text()
+        return f"""
+            You are a legal information extraction engine.
```
Comment (longhotsummer) [inline peachjam/analysis/criminal_data/offence.py]: This is the system or developer prompt. The LLM gives prompts with "role: system" (can only be one, right at the start) or "role: developer" special consideration.

Then provide the document text as "role: user". That helps to separate instructions from content.

Comment (longhotsummer) [reply peachjam/analysis/criminal_data/offence.py]: See https://github.com/laws-africa/peachjam/blob/0368f36ba3e6a18ed23a0d26ab639e5f8fd8bc8b/peachjam/chat/agent.py#L123

File: peachjam/analysis/criminal_data/base.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        Call the LLM service
+        """
+        prompt = self.build_prompt()
+
+        response = self.client.responses.parse(
+            model=self.model,
+            input=[
+                {"role": "system", "content": self.system_prompt()},
+                {"role": "user", "content": prompt},
```
Comment (longhotsummer) [inline peachjam/analysis/criminal_data/base.py]: I've found it useful to be able to pass multiple messages as input, using the different roles available. 

Use `instructions=self.system_prompt()` then `input=self.build_prompt()` which will allow `build_prompt()` to return a list of messages, rather than one, which is useful.

File: peachjam/analysis/criminal_data/offence_tool.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    def tools(self):
+        return [SEARCH_OFFENCES_TOOL]
+
+    def tool_handlers(self):
+        return {"search_offences": search_offences}
+
+    def build_prompt(self) -> str:
+        judgment_text = self.judgment.get_content_as_text()
+        return f"""
+            <ROLE>
```
Comment (longhotsummer) [inline peachjam/analysis/criminal_data/offence_tool.py]: make this your system prompt

File: peachjam/analysis/criminal_data/agent.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+log = logging.getLogger(__name__)
+
+
+KeywordInput = Union[str, Sequence[str]]
+_SPLIT = re.compile(r"[,\n;|]+")
+
+
+@function_tool
+def search_offences(
+    keywords: KeywordInput,
```
Comment (longhotsummer) [inline peachjam/analysis/criminal_data/agent.py]: I suggest to avoid being ambiguous: either make this a single string that is comma-separated, or a list of strings, not both. Don't make the LLM make a choice it doesn't have to.

File: peachjam/tests/test_criminal_data_extraction.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            title="Criminal trespass",
+            description="""
+            A person who unlawfully enters into or upon property in the possession of another with "
+            intent to commit an offence or to intimidate, insult or annoy any person in possession of
+            the property, or who, having lawfully entered, unlawfully remains there with such intent,
+            commits criminal trespass.
+            """,
+        )
+
+    def test_extract_robbery_with_violence(self):
```
Comment (longhotsummer) [inline peachjam/tests/test_criminal_data_extraction.py]: great, much simpler!

You need some tests that cover other edge cases for the fields in your pydantic models, such as "minimum sentence applied", or poorly matching tool results, or ambiguous judgment text where other sentences are mentioned that aren't the actual sentence.

File: peachjam/analysis/criminal_data/agent.py (hunk 4 above)
Comment (actlikewill) [reply peachjam/analysis/criminal_data/agent.py]: i've removed the config parameters from function call

File: peachjam/signals.py [hunk 6]
```diff
@@ -226,6 +228,26 @@ def judgment_content_changed_generate_summary(sender, instance, **kwargs):
         generate_judgment_summary(judgment.pk)
 
 
+@receiver(signals.pre_save, sender=DocumentContent)
+def judgment_content_changed_extract_criminal_data(sender, instance, **kwargs):
```
Comment (longhotsummer) [inline peachjam/signals.py]: I think this should check if criminal data extraction is enabled, and do nothing if it is not. Otherwise it's going to queue up background tasks which fail due to the ValueError raised when the extractor is created.

I think having the check here is better than the extractor's constructor raising an exception. I think of this as the settings flag controlling whether the work is queued up to be done, rather than the actual analysis logic itself checking if it's allowed to be run.
